File: python/src/envdoctor/scanner.py
"""Core environment-variable consistency scanner (native Python port).

Local-first: reads ``.env`` files and scans Python source for environment
access, then reconciles the two. No network, no telemetry, values never printed.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable

# Source usage patterns. Each captures the variable name in group 1.
_USAGE_PATTERNS = [
    re.compile(r"\bos\.environ\.get\(\s*[\"']([A-Za-z_]\w*)[\"']"),
    re.compile(r"\bos\.getenv\(\s*[\"']([A-Za-z_]\w*)[\"']"),
    re.compile(r"\bos\.environ\[\s*[\"']([A-Za-z_]\w*)[\"']\s*\]"),
    re.compile(r"\benviron\.get\(\s*[\"']([A-Za-z_]\w*)[\"']"),
    re.compile(r"\benviron\[\s*[\"']([A-Za-z_]\w*)[\"']\s*\]"),
]

# Triple-quoted docstrings are stripped so example code inside them is ignored.
_DOCSTRING = re.compile(r"\"\"\".*?\"\"\"|'''.*?'''", re.DOTALL)
_LINE_COMMENT = re.compile(r"#[^\n]*")

_ENV_LINE = re.compile(r"^\s*(?:export\s+)?([A-Za-z_]\w*)\s*=(.*)$")

# Public/client-exposed environment prefixes (case-sensitive, exact).
_PUBLIC_PREFIXES = (
    "NEXT_PUBLIC_",
    "VITE_",
    "REACT_APP_",
    "EXPO_PUBLIC_",
    "GATSBY_",
    "NUXT_PUBLIC_",
    "VUE_APP_",
    "PUBLIC_",
)

# Secret-looking name pattern (case-insensitive).
_SECRET_NAME = re.compile(
    r"SECRET|TOKEN|PASSWORD|PASSWD|PRIVATE|CREDENTIAL|API_?KEY|ACCESS_?KEY|AUTH",
    re.IGNORECASE,
)

# Weak/placeholder secret value pattern (case-insensitive, whole-value).
_WEAK_SECRET_VALUE = re.compile(
    r"^(changeme|change_me|placeholder|x{3,}|todo|secret|password|passwd|test|"
    r"example|sample|dummy|your[_-].*|<.*>|\$\{.*\})$",
    re.IGNORECASE,
)

_INT_RE = re.compile(r"^-?\d+$")
_FLOAT_RE = re.compile(r"^-?\d+\.\d+$")
_URL_RE = re.compile(r"^https?://", re.IGNORECASE)


@dataclass
class Origin:
    file: Path
    line: int


@dataclass
class Finding:
    rule: str
    severity: str  # "error" | "warning"
    name: str
    message: str
    origin: Origin | None = None


@dataclass
class ScanResult:
    findings: list[Finding] = field(default_factory=list)

    @property
    def errors(self) -> list[Finding]:
        return [f for f in self.findings if f.severity == "error"]

    @property
    def warnings(self) -> list[Finding]:
        return [f for f in self.findings if f.severity == "warning"]


def _strip_noise(code: str) -> str:
    """Blank docstrings and comments while preserving line structure."""

    def blank(match: re.Match[str]) -> str:
        return re.sub(r"[^\n]", " ", match.group(0))

    code = _DOCSTRING.sub(blank, code)
    code = _LINE_COMMENT.sub(blank, code)
    return code


def parse_value(raw_rhs: str) -> str:
    """Trim the right-hand side and strip one matching pair of surrounding quotes."""
    v = raw_rhs.strip()
    if len(v) >= 2 and v[0] == v[-1] and v[0] in ("'", '"'):
        v = v[1:-1]
    return v


def env_label(filename: str) -> str:
    """Derive an environment label from a dotenv filename."""
    if filename == ".env":
        return "default"
    # filename starts with ".env." here (discover filters accordingly).
    rest = filename[len(".env."):]
    if rest == "local":
        return "local"
    if rest.endswith(".local"):
        rest = rest[: -len(".local")]
    return rest


def infer_type(value: str) -> str:
    """Infer a coarse type from a value string."""
    if value == "":
        return "empty"
    if _INT_RE.match(value):
        return "integer"
    if _FLOAT_RE.match(value):
        return "float"
    if value.lower() in ("true", "false"):
        return "boolean"
    if _URL_RE.match(value):
        return "url"
    if value[0] in ("{", "["):
        try:
            json.loads(value)
            return "json"
        except ValueError:
            pass
    return "string"


def compat_group(inferred: str) -> str:
    """Map an inferred type to its compatibility group."""
    if inferred in ("integer", "float"):
        return "numeric"
    return inferred


def levenshtein(a: str, b: str) -> int:
    """Classic dynamic-programming Levenshtein edit distance."""
    if a == b:
        return 0
    if not a:
        return len(b)
    if not b:
        return len(a)
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a, start=1):
        cur = [i]
        for j, cb in enumerate(b, start=1):
            cost = 0 if ca == cb else 1
            cur.append(min(prev[j] + 1, cur[j - 1] + 1, prev[j - 1] + cost))
        prev = cur
    return prev[-1]


@dataclass
class _Def:
    origin: Origin
    values: dict[str, str] = field(default_factory=dict)  # label -> value


def parse_env_file(path: Path) -> dict[str, list[tuple[Origin, str]]]:
    """Return ``{NAME: [(Origin, value), ...]}`` for every definition in a file."""
    defined: dict[str, list[tuple[Origin, str]]] = {}
    for lineno, raw in enumerate(path.read_text().splitlines(), start=1):
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        match = _ENV_LINE.match(raw)
        if match:
            value = parse_value(match.group(2))
            defined.setdefault(match.group(1), []).append((Origin(path, lineno), value))
    return defined


def scan_source_file(path: Path) -> dict[str, Origin]:
    """Return ``{NAME: first Origin}`` for env usages in a Python source file."""
    text = _strip_noise(path.read_text())
    used: dict[str, Origin] = {}
    for pattern in _USAGE_PATTERNS:
        for match in pattern.finditer(text):
            name = match.group(1)
            line = text.count("\n", 0, match.start()) + 1
            used.setdefault(name, Origin(path, line))
    return used


def discover_env_files(root: Path) -> list[Path]:
    files = sorted(root.glob(".env"))
    files += sorted(p for p in root.glob(".env.*") if not p.name.endswith(".example"))
    return files


def defined_by_label(root: Path) -> dict[str, set[str]]:
    """Map each environment label to the set of variable names defined in it."""
    labels: dict[str, set[str]] = {}
    for env_file in discover_env_files(root):
        names = labels.setdefault(env_label(env_file.name), set())
        for name in parse_env_file(env_file):
            names.add(name)
    return labels


def diff_labels(root: Path, a: str, b: str) -> dict[str, list[str]]:
    """Compare the variable names defined in two environment labels."""
    labels = defined_by_label(root)
    da = labels.get(a, set())
    db = labels.get(b, set())
    return {
        "onlyInA": sorted(da - db),
        "onlyInB": sorted(db - da),
        "common": sorted(da & db),
    }


def sync_labels(root: Path, src: str, dst: str, dry_run: bool = False) -> list[str]:
    """Append keys present in ``src`` but missing from ``dst`` to dst's file.

    Values are never copied — only ``KEY=`` placeholders are written.
    """
    labels = defined_by_label(root)
    missing = sorted(labels.get(src, set()) - labels.get(dst, set()))
    if missing and not dry_run:
        target = root / (".env" if dst == "default" else f".env.{dst}")
        existing = target.read_text() if target.exists() else ""
        prefix = "" if existing == "" or existing.endswith("\n") else "\n"
        with target.open("a") as fh:
            fh.write(prefix + "".join(f"{k}=\n" for k in missing))
    return missing


def discover_source_files(root: Path, extensions: Iterable[str] = ("py",)) -> list[Path]:
    exts = {e.lstrip(".").lower() for e in extensions}
    out: list[Path] = []
    for path in sorted(root.rglob("*")):
        if not path.is_file():
            continue
        if any(part in {".git", "__pycache__", ".venv", "node_modules"} for part in path.parts):
            continue
        if path.suffix.lstrip(".").lower() in exts:
            out.append(path)
    return out


def scan(root: Path, extensions: Iterable[str] = ("py",)) -> ScanResult:
    """Reconcile dotenv definitions against source usage under ``root``."""
    defined: dict[str, _Def] = {}
    project_labels: set[str] = set()
    duplicates: list[Finding] = []

    for env_file in discover_env_files(root):
        label = env_label(env_file.name)
        project_labels.add(label)
        for name, occurrences in parse_env_file(env_file).items():
            first_origin, first_value = occurrences[0]
            if name not in defined:
                defined[name] = _Def(origin=first_origin)
            # First value seen for this label wins.
            if label not in defined[name].values:
                defined[name].values[label] = first_value
            # Duplicate within a single file: 2+ occurrences of the same key.
            if len(occurrences) >= 2:
                lines = ", ".join(str(o.line) for o, _ in occurrences)
                duplicates.append(
                    Finding(
                        rule="duplicates",
                        severity="error",
                        name=name,
                        message=(
                            f"defined {len(occurrences)} times in the same file "
                            f"(lines {lines})"
                        ),
                        origin=first_origin,
                    )
                )

    used: dict[str, Origin] = {}
    for src in discover_source_files(root, extensions):
        for name, origin in scan_source_file(src).items():
            used.setdefault(name, origin)

    result = ScanResult()

    # --- Errors, in rule order ---

    # undefined-in-source: used but never defined.
    for name in sorted(used):
        if name not in defined:
            result.findings.append(
                Finding(
                    rule="undefined-in-source",
                    severity="error",
                    name=name,
                    message="used in source code but not defined in any environment file",
                    origin=used[name],
                )
            )

    # duplicates: same key defined 2+ times within a single .env file.
    for finding in sorted(duplicates, key=lambda f: f.name):
        result.findings.append(finding)

    # public-prefix: secret-looking var exposed to client bundles.
    for name in sorted(defined):
        if name.startswith(_PUBLIC_PREFIXES) and _SECRET_NAME.search(name):
            result.findings.append(
                Finding(
                    rule="public-prefix",
                    severity="error",
                    name=name,
                    message=(
                        "secret-looking variable is exposed to client bundles "
                        "via a public prefix"
                    ),
                    origin=defined[name].origin,
                )
            )

    # type-mismatch: incompatible inferred types across environments.
    for name in sorted(defined):
        d = defined[name]
        if len(d.values) < 2:
            continue
        groups = set()
        for value in d.values.values():
            t = infer_type(value)
            if t == "empty":
                continue
            groups.add(compat_group(t))
        if len(groups) >= 2:
            result.findings.append(
                Finding(
                    rule="type-mismatch",
                    severity="error",
                    name=name,
                    message="inferred type differs across environments",
                    origin=d.origin,
                )
            )

    # --- Warnings, in rule order ---

    # unused: defined but never referenced.
    for name in sorted(defined):
        if name not in used:
            result.findings.append(
                Finding(
                    rule="unused",
                    severity="warning",
                    name=name,
                    message="defined but never referenced in source",
                    origin=defined[name].origin,
                )
            )

    # environment-diff: defined in some but not all project env labels.
    if len(project_labels) >= 2:
        for name in sorted(defined):
            present = sorted(defined[name].values)
            absent = sorted(project_labels - set(present))
            if present and absent:
                result.findings.append(
                    Finding(
                        rule="environment-diff",
                        severity="warning",
                        name=name,
                        message=(
                            f"defined in {', '.join(present)} but missing in "
                            f"{', '.join(absent)}"
                        ),
                        origin=defined[name].origin,
                    )
                )

    # weak-secret: secret-looking name with a weak or placeholder value.
    for name in sorted(defined):
        if not _SECRET_NAME.search(name):
            continue
        value = defined[name].values.get(env_label(defined[name].origin.file.name))
        if value is None:
            value = next(iter(defined[name].values.values()), "")
        if value == "" or len(value) < 8 or _WEAK_SECRET_VALUE.match(value):
            result.findings.append(
                Finding(
                    rule="weak-secret",
                    severity="warning",
                    name=name,
                    message="secret-looking variable has a weak or placeholder value",
                    origin=defined[name].origin,
                )
            )

    # typo: used-but-undefined name close to a defined name.
    defined_names = list(defined)
    for u in sorted(used):
        if u in defined:
            continue
        best_d: str | None = None
        best_dist = 0
        for d in sorted(defined_names):
            if d == u:
                continue
            threshold = 1 if min(len(u), len(d)) <= 4 else 2
            dist = levenshtein(u, d)
            if dist <= threshold and (best_d is None or dist < best_dist):
                best_d, best_dist = d, dist
        if best_d is not None:
            result.findings.append(
                Finding(
                    rule="typo",
                    severity="warning",
                    name=u,
                    message=f'"{u}" is not defined; did you mean "{best_d}"?',
                    origin=used[u],
                )
            )

    return result


def finding_to_dict(finding: Finding, root: Path) -> dict:
    """Serialize a finding for JSON output. Values are never included."""
    file = None
    line = None
    if finding.origin is not None:
        try:
            file = str(finding.origin.file.relative_to(root))
        except ValueError:
            file = str(finding.origin.file)
        line = finding.origin.line
    return {
        "rule": finding.rule,
        "severity": finding.severity,
        "name": finding.name,
        "message": finding.message,
        "file": file,
        "line": line,
    }

File: ruby/lib/envdoctor/scanner.rb
# frozen_string_literal: true

require "json"

module Envdoctor
  # Core scanner: reconcile ENV usage in Ruby source against .env definitions.
  # Local-first — no network, values never printed.
  module Scanner
    module_function

    USAGE_PATTERNS = [
      /\bENV\[\s*["']([A-Za-z_]\w*)["']\s*\]/,
      /\bENV\.fetch\(\s*["']([A-Za-z_]\w*)["']/
    ].freeze

    ENV_LINE = /\A\s*(?:export\s+)?([A-Za-z_]\w*)\s*=/.freeze

    PUBLIC_PREFIXES = %w[
      NEXT_PUBLIC_ VITE_ REACT_APP_ EXPO_PUBLIC_ GATSBY_ NUXT_PUBLIC_ VUE_APP_ PUBLIC_
    ].freeze

    SECRET_RE = /SECRET|TOKEN|PASSWORD|PASSWD|PRIVATE|CREDENTIAL|API_?KEY|ACCESS_?KEY|AUTH/i.freeze

    WEAK_VALUE_RE =
      /\A(changeme|change_me|placeholder|x{3,}|todo|secret|password|passwd|test|example|sample|dummy|your[_-].*|<.*>|\$\{.*\})\z/i.freeze

    Origin = Struct.new(:file, :line)
    # A single .env definition occurrence: its line and parsed value.
    Definition = Struct.new(:line, :value)
    Finding = Struct.new(:rule, :severity, :name, :message, :origin)

    # Blank comments and =begin/=end blocks, preserving line structure.
    def strip_noise(code)
      code = code.gsub(/^=begin\b.*?^=end\b[^\n]*/m) { |m| m.gsub(/[^\n]/, " ") }
      code.gsub(/#[^\n]*/) { |m| " " * m.length }
    end

    def scan_source(path, content)
      text = strip_noise(content)
      used = {}
      USAGE_PATTERNS.each do |re|
        text.to_enum(:scan, re).each do
          match = Regexp.last_match
          name = match[1]
          next if used.key?(name)

          line = text[0...match.begin(0)].count("\n") + 1
          used[name] = Origin.new(path, line)
        end
      end
      used
    end

    # Parse the VALUE to the right of the first `=`: trim, then strip one pair
    # of matching surrounding quotes. Values are used ONLY for detection and are
    # never surfaced in any output.
    def parse_value(raw)
      idx = raw.index("=")
      return "" if idx.nil?

      value = raw[(idx + 1)..].to_s.strip
      if value.length >= 2 && %w[" '].include?(value[0]) && value[-1] == value[0]
        value = value[1..-2]
      end
      value
    end

    # Returns { name => [Definition, ...] } with ALL occurrences per key in order.
    def parse_env(path, content)
      defined = {}
      content.split("\n").each_with_index do |raw, i|
        stripped = raw.strip
        next if stripped.empty? || stripped.start_with?("#")

        if (m = raw.match(ENV_LINE))
          (defined[m[1]] ||= []) << Definition.new(i + 1, parse_value(raw))
        end
      end
      defined
    end

    # Derive the environment label from a .env filename.
    def env_label(filename)
      base = File.basename(filename)
      return "default" if base == ".env"

      label = base.sub(/\A\.env\./, "")
      label = label.sub(/\.local\z/, "") if label.end_with?(".local")
      label
    end

    def public_prefix?(name)
      PUBLIC_PREFIXES.any? { |p| name.start_with?(p) } && SECRET_RE.match?(name)
    end

    # Infer the coarse type of a value string.
    def infer_type(value)
      return "empty" if value.empty?
      return "integer" if value.match?(/\A-?\d+\z/)
      return "float" if value.match?(/\A-?\d+\.\d+\z/)
      return "boolean" if value.match?(/\A(true|false)\z/i)
      return "url" if value.match?(%r{\Ahttps?://})

      if value.start_with?("{", "[")
        begin
          JSON.parse(value)
          return "json"
        rescue JSON::ParserError
          # fall through to string
        end
      end
      "string"
    end

    # Compatibility group for an inferred type (integer/float collapse to numeric).
    def type_group(type)
      %w[integer float].include?(type) ? "numeric" : type
    end

    def weak_secret?(value)
      value.empty? || value.length < 8 || WEAK_VALUE_RE.match?(value)
    end

    def levenshtein(a, b)
      return b.length if a.empty?
      return a.length if b.empty?

      prev = (0..b.length).to_a
      a.each_char.with_index do |ca, i|
        curr = [i + 1]
        b.each_char.with_index do |cb, j|
          cost = ca == cb ? 0 : 1
          curr << [curr[j] + 1, prev[j + 1] + 1, prev[j] + cost].min
        end
        prev = curr
      end
      prev[b.length]
    end

    def discover_env_files(root)
      files = Dir.glob(File.join(root, ".env"))
      files += Dir.glob(File.join(root, ".env.*")).reject { |f| f.end_with?(".example") }
      files.sort
    end

    def discover_source_files(root)
      Dir.glob(File.join(root, "**", "*.rb")).reject do |p|
        p.split(File::SEPARATOR).any? { |part| %w[.git vendor node_modules].include?(part) }
      end.sort
    end

    COMPOSE_RE = /\A(docker-)?compose([.-].*)?\.ya?ml\z/i.freeze

    def skip_infra_path?(path)
      path.split(File::SEPARATOR).any? { |part| %w[.git vendor node_modules target].include?(part) }
    end

    # Walk the project and classify infra YAML files. Returns [[path, kind], ...]
    # where kind is one of :compose, :actions, :k8s.
    def discover_infra_files(root)
      out = []
      Dir.glob(File.join(root, "**", "*"), File::FNM_DOTMATCH).each do |path|
        next unless File.file?(path)
        next if skip_infra_path?(path)

        base = File.basename(path)
        segments = path.split(File::SEPARATOR)
        if base.match?(COMPOSE_RE)
          out << [path, :compose]
        elsif base =~ /\.ya?ml\z/i &&
              segments.each_cons(2).any? { |a, b| a == ".github" && b == "workflows" }
          out << [path, :actions]
        elsif base =~ /\.ya?ml\z/i
          content = File.read(path)
          out << [path, :k8s] if content.match?(/^apiVersion:/m) && content.match?(/^kind:/m)
        end
      end
      out.sort_by(&:first)
    end

    INTERP_PATTERNS = [
      /\$\{([A-Za-z_][A-Za-z0-9_]*)/,
      /\$([A-Za-z_][A-Za-z0-9_]*)/
    ].freeze

    ACTIONS_CONTEXT_RE = /\b(?:secrets|vars|env)\.([A-Za-z_][A-Za-z0-9_]*)/.freeze

    # Extract used variable NAMES (with origin) from an infra YAML file. First
    # removes escaped `$$` (two spaces, offsets preserved), then scans for
    # interpolation and, for GitHub Actions, the secrets/vars/env contexts.
    def scan_infra(path, content, kind)
      text = content.gsub("$$", "  ")
      used = {}
      patterns = INTERP_PATTERNS.dup
      patterns << ACTIONS_CONTEXT_RE if kind == :actions
      patterns.each do |re|
        text.to_enum(:scan, re).each do
          match = Regexp.last_match
          name = match[1]
          next if used.key?(name)

          line = text[0...match.begin(0)].count("\n") + 1
          used[name] = Origin.new(path, line)
        end
      end
      used
    end

    # Map each environment label to the set of variable names defined in it.
    def defined_by_label(root)
      labels = {}
      discover_env_files(root).each do |f|
        set = (labels[env_label(f)] ||= [])
        parse_env(f, File.read(f)).each_key { |name| set << name unless set.include?(name) }
      end
      labels
    end

    def diff_labels(root, a, b)
      labels = defined_by_label(root)
      da = labels[a] || []
      db = labels[b] || []
      {
        "onlyInA" => (da - db).sort,
        "onlyInB" => (db - da).sort,
        "common" => (da & db).sort
      }
    end

    # Append keys present in `from` but missing from `to` as `KEY=` placeholders.
    # Values are never copied.
    def sync_labels(root, from, to, dry_run: false)
      labels = defined_by_label(root)
      missing = ((labels[from] || []) - (labels[to] || [])).sort
      if !missing.empty? && !dry_run
        target = File.join(root, to == "default" ? ".env" : ".env.#{to}")
        existing = File.exist?(target) ? File.read(target) : ""
        prefix = (existing.empty? || existing.end_with?("\n")) ? "" : "\n"
        File.open(target, "a") { |fh| fh.write(prefix + missing.map { |k| "#{k}=\n" }.join) }
      end
      missing
    end

    # Collect the DEFINED (from .env*) and USED (source + infra) name sets for a
    # project. Returns [defined_hash, used_hash] with name => true entries.
    def collect_names(root)
      defined = {}
      discover_env_files(root).each do |f|
        parse_env(relative(root, f), File.read(f)).each_key { |name| defined[name] = true }
      end
      used = {}
      discover_source_files(root).each do |f|
        scan_source(relative(root, f), File.read(f)).each_key { |k| used[k] = true }
      end
      discover_infra_files(root).each do |f, kind|
        scan_infra(relative(root, f), File.read(f), kind).each_key { |k| used[k] = true }
      end
      [defined, used]
    end

    # Exact `.env.example` content: header, then `NAME=` per (defined ∪ used),
    # sorted ascending. Values are never written.
    def env_example_content(root)
      defined, used = collect_names(root)
      names = (defined.keys + used.keys).uniq.sort
      "# Generated by envdoctor. Fill in values; do not commit secrets.\n" +
        names.map { |n| "#{n}=\n" }.join
    end

    # Exact `ENVIRONMENT.md` content: a table of Defined/Used per variable.
    def environment_doc_content(root)
      defined, used = collect_names(root)
      names = (defined.keys + used.keys).uniq.sort
      lines = ["# Environment variables", "", "| Variable | Defined | Used |", "| --- | --- | --- |"]
      names.each do |n|
        lines << "| #{n} | #{defined.key?(n) ? 'yes' : 'no'} | #{used.key?(n) ? 'yes' : 'no'} |"
      end
      lines.join("\n") + "\n"
    end

    def scan(root)
      defined = {}          # name => Origin (first definition wins)
      defined_value = {}    # name => value at first definition
      labels_of = {}        # name => { label => value } (first value per label)
      project_labels = []
      dup_findings = []

      discover_env_files(root).each do |f|
        rel = relative(root, f)
        label = env_label(f)
        project_labels << label unless project_labels.include?(label)
        parse_env(rel, File.read(f)).each do |name, defs|
          if defs.length >= 2
            lines = defs.map(&:line)
            dup_findings << Finding.new("duplicates", "error", name,
                                        "defined #{defs.length} times in the same file " \
                                        "(lines #{lines.join(', ')})",
                                        Origin.new(rel, defs.first.line))
          end
          # First occurrence (first file wins) counts as the definition.
          unless defined.key?(name)
            defined[name] = Origin.new(rel, defs.first.line)
            defined_value[name] = defs.first.value
          end
          bucket = (labels_of[name] ||= {})
          bucket[label] = defs.first.value unless bucket.key?(label)
        end
      end

      used = {}
      discover_source_files(root).each do |f|
        scan_source(relative(root, f), File.read(f)).each { |k, v| used[k] ||= v }
      end
      # Infra sources (Docker Compose / GitHub Actions / Kubernetes) contribute
      # additional used names; first origin wins.
      discover_infra_files(root).each do |f, kind|
        scan_infra(relative(root, f), File.read(f), kind).each { |k, v| used[k] ||= v }
      end

      errors = []
      warnings = []

      # --- errors: undefined-in-source ---
      used.keys.sort.each do |name|
        next if defined.key?(name)

        errors << Finding.new("undefined-in-source", "error", name,
                              "referenced but not defined in any environment file",
                              used[name])
      end

      # --- errors: duplicates ---
      dup_findings.sort_by(&:name).each { |finding| errors << finding }

      # --- errors: public-prefix ---
      defined.keys.sort.each do |name|
        next unless public_prefix?(name)

        errors << Finding.new("public-prefix", "error", name,
                              "secret-looking variable is exposed to client bundles " \
                              "via a public prefix", defined[name])
      end

      # --- errors: type-mismatch ---
      defined.keys.sort.each do |name|
        labels = labels_of[name] || {}
        next if labels.size < 2

        groups = labels.values.map { |v| infer_type(v) }.reject { |t| t == "empty" }
                       .map { |t| type_group(t) }.uniq
        next if groups.size < 2

        errors << Finding.new("type-mismatch", "error", name,
                              "inferred type differs across environments", defined[name])
      end

      # --- errors: schema-validation ---
      load_schema(root).sort.each do |name, rule|
        next unless rule.is_a?(Hash)

        if defined.key?(name)
          msg = schema_failure(rule, defined_value[name])
          errors << Finding.new("schema-validation", "error", name, msg, defined[name]) if msg
        elsif !rule["optional"]
          errors << Finding.new("schema-validation", "error", name,
                                "required by schema but not defined", nil)
        end
      end

      # --- warnings: unused ---
      defined.keys.sort.each do |name|
        next if used.key?(name)

        warnings << Finding.new("unused", "warning", name,
                                "defined but never referenced in source", defined[name])
      end

      # --- warnings: environment-diff ---
      if project_labels.length >= 2
        defined.keys.sort.each do |name|
          present = (labels_of[name] || {}).keys.sort
          absent = (project_labels - present).sort
          next if present.empty? || absent.empty?

          warnings << Finding.new("environment-diff", "warning", name,
                                  "defined in #{present.join(', ')} but missing in " \
                                  "#{absent.join(', ')}", defined[name])
        end
      end

      # --- warnings: weak-secret ---
      defined.keys.sort.each do |name|
        next unless SECRET_RE.match?(name)
        next unless weak_secret?(defined_value[name].to_s)

        warnings << Finding.new("weak-secret", "warning", name,
                                "secret-looking variable has a weak or placeholder value",
                                defined[name])
      end

      # --- warnings: typo ---
      defined_names = defined.keys
      used.keys.sort.each do |u|
        next if defined.key?(u)

        best = nil
        best_dist = nil
        defined_names.each do |d|
          next if d == u

          limit = [u.length, d.length].min <= 4 ? 1 : 2
          dist = levenshtein(u, d)
          next if dist > limit

          if best.nil? || dist < best_dist || (dist == best_dist && d < best)
            best = d
            best_dist = dist
          end
        end
        next if best.nil?

        warnings << Finding.new("typo", "warning", u,
                                "\"#{u}\" is not defined; did you mean \"#{best}\"?",
                                used[u])
      end

      errors + warnings
    end

    # Serialize findings to the shared JSON shape. Values never appear.
    NUMERIC_RE = /\A-?\d+(\.\d+)?\z/.freeze

    def load_schema(root)
      require "json"
      data = JSON.parse(File.read(File.join(root, "envdoctor.schema.json")))
      data.is_a?(Hash) ? data : {}
    rescue StandardError
      {}
    end

    def schema_type_ok(value, declared)
      case declared
      when "string" then true
      when "integer" then value.match?(/\A-?\d+\z/)
      when "float" then value.match?(/\A-?\d+(\.\d+)?\z/)
      when "boolean" then %w[true false].include?(value.downcase)
      when "url" then value.match?(%r{\Ahttps?://})
      when "json" then (JSON.parse(value); true rescue false)
      else true
      end
    end

    def schema_failure(rule, value)
      t = rule["type"]
      return "value does not match schema type #{t}" if t.is_a?(String) && !schema_type_ok(value, t)

      enum = rule["enum"]
      return "value is not one of the allowed values" if enum.is_a?(Array) && !enum.include?(value)

      pattern = rule["regex"]
      if pattern.is_a?(String)
        begin
          return "value does not match the required pattern" if value !~ Regexp.new(pattern)
        rescue RegexpError
          # ignore invalid pattern
        end
      end

      if value.match?(NUMERIC_RE)
        num = value.to_f
        lo = rule["min"]
        return "value is below the minimum" if lo.is_a?(Numeric) && num < lo

        hi = rule["max"]
        return "value exceeds the maximum" if hi.is_a?(Numeric) && num > hi
      end
      nil
    end

    def to_json_array(findings)
      JSON.generate(findings.map do |f|
        {
          "rule" => f.rule,
          "severity" => f.severity,
          "name" => f.name,
          "message" => f.message,
          "file" => f.origin&.file,
          "line" => f.origin&.line
        }
      end)
    end

    def relative(root, path)
      path.sub(/\A#{Regexp.escape(root)}#{Regexp.escape(File::SEPARATOR)}?/, "")
    end
  end
end
